tuist/tuist
# [DRAFT] Add `use_default_xcode` option to `fourier bundle` command

> ### This PR is not ready for prime-time. I'm opening this as a Draft to get discussion around the idea started.
> 
> I'd appreciate any feedback on my approach here, and the idea itself. This is my first time making a real change to `fourier`, so I'm not familiar with the conventions here. I know I'm probably breaking some rules. 😅

### Short description 📝

The changes in #3940 make the `fourier bundle` command explicitly set the Xcode version to be used when building the `tuist` binary and its frameworks. This makes total sense in the context of CI, and fixes the problems the PR was opened to address, but for other uses this isn't ideal. I like having the ability to build a `tuist` binary from source, and bundle that in our repo, but I'd prefer not to have to have multiple versions of Xcode on my machine in order to do so.

This is useful in the case that I need to use a feature or bug fix that has not yet been officially released. In this case, I'd build the project locally, and move it into our repo's `.tuist-bin` directory.

And so, I've added an optional flag to the `fourier bundle` command: `--use_default_xcode` or `-d`. This skips over the code that sets the Xcode version during the execution of `fourier bundle`, thereby using whatever version of Xcode is currently selected on the machine doing the build.

There are potential compatibility issues when using this flag — which #3940 set out to correct — so we would need to document those if we adopt this change (or something like it). In an environment like the one I work in for my day-to-day, the version of Xcode and macOS that our developers use is tightly controlled, so it's not a concern, but obviously that's not the case for everyone.

#### Alternatives Considered

- A user could, of course, modify the `.xcode-version` and `.xcode-version-libraries` files to their desired version number prior to bundling Tuist with `fourier`.
    - This option makes it more difficult to run the command from a custom script.
    - Also, the current implementation expects the name of the Xcode app bundle to be in a specific format. This isn't a problem on Tuist's CI, but in our environment, we name Xcode slightly differently, so it would require changing the name of Xcode on our machines. (`fourier` currently expects `Xcode_13.2.1.app`, whereas we use `Xcode-13.2.1.app` — the default naming done by the [`xcodes`](https://github.com/RobotsAndPencils/xcodes) tool we use to download Xcode versions.)
- We could instead add an `--xcode-version` argument, or `--xcode-version` and `--xcode-version-libraries` arguments, that would allow us to specify the version needed from the command line. This would be a little more complex solution, but may be better in the long run.
    - With this option, we could also consider removing the `.xcode-version` and `.xcode-version-libraries` files from the repo. I'm not sure if that's worth doing, but just an option.

### How to test the changes locally 🧐

1. Run `fourier bundle tuist -d` and observe that the binary is built with whatever version of Xcode is currently selected on your machine.
